### PR TITLE
Add support for Dremio as a new source

### PR DESCRIPTION
**Pull request by narendrans**
### CATEGORY

 Enhancement (new features, refinement)
 

### SUMMARY

Added support to connect to Dremio. Tested it on the master.

### ADDITIONAL INFORMATION

- [ ] Has associated issue:
- [ ] Changes UI
- [ ] Requires DB Migration.
- [ ] Confirm DB Migration upgrade and downgrade tested.
- [ ] Introduces new feature or API
- [ ] Removes existing feature or API

### CODE REVIEW

**Comment by narendrans** [comment]
Is there a way to add the required import `sqlalchemy_dremio.pyodbc` somewhere in requirements.txt?  I have added that in the `db_engine_specs/dremio.py`.

**Comment by villebro** [comment]
@narendrans you can add the optional requirements in`setup.py`

File: superset/db_engine_specs/dremio.py [hunk 1]
```diff
@@ -0,0 +1,18 @@
...
+        None: "{col}",
+        "PT1S": "DATE_TRUNC('second', {col})",
+        "PT1M": "DATE_TRUNC('minute', {col})",
+        "PT1H": "DATE_TRUNC('hour', {col})",
+        "P1D": "DATE_TRUNC('day', {col})",
+        "P1W": "DATE_TRUNC('week', {col})",
+        "P1M": "DATE_TRUNC('month', {col})",
+        "P0.25Y": "DATE_TRUNC('quarter', {col})",
+        "P1Y": "DATE_TRUNC('year', {col})",
+    }
```
Comment (villebro) [inline superset/db_engine_specs/dremio.py]: Please add line change to last row.

Comment (narendrans) [reply superset/db_engine_specs/dremio.py]: I don't get you. Can you please rephrase?

Comment (narendrans) [reply superset/db_engine_specs/dremio.py]: > @narendrans you can add the optional requirements in`setup.py`

Done.

**Comment by narendrans** [comment]
@villebro Thanks for reviewing the PR. Let me know if it looks good with my recent changes.

File: superset/db_engine_specs/dremio.py (hunk 1 above)
Comment (narendrans) [reply superset/db_engine_specs/dremio.py]: It looks like I still need to manually add `import sqlalchemy_dremio` in dremio.py. How does superset automatically import the  one in `extras_require={`?

**Comment by villebro** [comment]
Quick second pass:
- you need to add ASF license headers to the new file.
- you need to run `isort` (import order incorrect).
- you need to run `black` (reformatting).
- you need to implement `epoch_to_dttm()` in the the engine spec.

You can check these errors in the CI log.

File: setup.py [hunk 2]
```diff
@@ -118,6 +118,7 @@ def get_git_sha():
         "elasticsearch": ["elasticsearch-dbapi>=0.1.0, <0.2.0"],
         "druid": ["pydruid==0.5.7", "requests==2.22.0"],
         "hana": ["hdbcli==2.4.162", "sqlalchemy_hana==0.4.0"],
+        "dremio": ["sqlalchemy_dremio"]
```
Comment (villebro) [inline setup.py]: We usually set the minimum version at the current most stable version (`>=0.5.0dev0`).

File: docs/installation.rst [hunk 3]
```diff
@@ -724,6 +726,15 @@ Druid
 Note that you can run the ``superset refresh_druid`` command to refresh the
 metadata from your Druid cluster(s)
 
+Dremio
+-----
+
+Install the following dependencies to connect to Dremio:
+
+* pyodbc `pip install pyodbc`
```
Comment (villebro) [inline docs/installation.rst]: I don't think this step is needed, as it's already a requirement in `sqlalchemy_dremio`: https://github.com/naren-dremio/sqlalchemy_dremio/blob/master/setup.py#L14-L18

**Review by villebro — CHANGES_REQUESTED** [review]
In addition to my comments earlier, a few more.

**Comment by narendrans** [comment]
@villebro I have made the changes as per your review. One final question:

It looks like I still need to manually add ``import sqlalchemy_dremio`` in ``dremio.py`` in my local test. How does superset automatically import the one in ``extras_require={``?

**Review by villebro — APPROVED** [review]
LGTM. Does this PR work with Dremio as-is, or are you getting exceptions for missing imports during use? If not, I think the connector might need some work. My quick first impression is that the modulemight not be registering correctly with SqlAlchemy. Normally SqlAlchemy should be able to pull in whatever imports it needs automatically. The only case where you should need to import the driver in the engine spec is if you're using functionality from said module. See e.g. https://github.com/apache/incubator-superset/blob/master/superset/db_engine_specs/hive.py#L86-L97 where the import is needed to catch the correct exception type.

**Comment by narendrans** [comment]
@villebro It looks like using the  URI ``dremio://`` prevents the issue. I have updated the docs to reflect that. Thanks for all your help!

**Comment by villebro** [comment]
Glad I could help and thanks for the PR (and by the looks of it the SqlAlchemy connector!)

**Comment by kunaldawn** [comment]
Hi, Is sql-lab working with dremio?
I had to patch : https://github.com/kunaldawn/foss-lake/blob/master/superset/dremio.py

**Comment by narendrans** [comment]
@kunaldawn Not sure if this is introduced in the new version. Are you referring to the issue where sql lab shows the results as arrays?

**Comment by kunaldawn** [comment]
> @kunaldawn Not sure if this is introduced in the new version. Are you referring to the issue where sql lab shows the results as arrays?

Yes exactly, I am referring  "sql lab shows the results as arrays". 

I think fetch_data is returning generator (also mentioned in https://github.com/apache/incubator-superset/issues/4192) (sorry, I am not very aware of superset code base, this is what I got between working sqlite vs non working dremio while debugging, and the data type need to be list of touple).

I referred the solution from https://github.com/apache/incubator-superset/pull/7048/files#diff-6519edc75f2440a575cb22492f401100R739, but that has list of list data type issue.

**Comment by villebro** [comment]
Can you open up an issue? I consider this a blocker for the `0.36` release, and it shouldn't be difficult to fix, but I would need to get some more context as I don't unfortunately have access to a Dremio instance.

**Comment by naren-dremio** [comment]
@villebro I assume this is only happening with Dremio? If so, I can send a PR with the changes proposed by @kunaldawn so that we can have it in the next release.

**Comment by villebro** [comment]
@naren-dremio the root cause of this problem was fixed last week, so this should now work without any additional changes to the dremio spec. Are you able to test the current master branch and help confirm that this is indeed the case?

**Comment by villebro** [comment]
@naren-dremio a dremio test instance would be very helpful, please reach out on Superset slack so we can set one up!

**Comment by narendrans** [comment]
@villebro Just tested it on master. Looks good. I will ping on slack so that we can have a test instance of Dremio.

**Pull request merged**